src/main/java/com/erigitic/jobs/TEAction.java

```diff
@@ -0,0 +1,196 @@
+                rewardMap.put(((String) k), reward);
+            });
+        }
+        return new TEAction(action, node.getKey().toString(), rewardMap, idTraitNode.getString("type"), growthTraitNode.getString(null));
+    }
+
+    private String action;
+    private String targetID;
+
+    /* Complex stuff if we need to evaluate traits */
+    private String growthTrait = null;
+    private String idTrait = null;
+    private Map<String, TEActionReward> rewards;
+
+    /* Simple stuff where we don't */
+    private TEActionReward reward;
+
+    private boolean isValid = false;
+
+    // Just construct an invalid one
+    protected TEAction() {}
+
+    public TEAction(String action, String targetID, TEActionReward reward, String growthTrait) {
+        this.action = action;
+        this.targetID = targetID;
+        this.growthTrait = growthTrait;
+        this.reward = reward;
+        isValid = (action != null && !action.isEmpty() && targetID != null && !targetID.isEmpty());
+    }
+
+    public TEAction(String action, String target, Map<String, TEActionReward> rewards, String idTrait, String growthTrait) {
+        this(action, target, null, growthTrait);
+        this.idTrait = idTrait;
+        this.rewards = rewards;
+    }
+
+    public boolean isGrowing() { return growthTrait != null; }
+
+    public boolean isIDTraited() { return idTrait != null; }
+
+    public String getIDTrait() { return idTrait; }
+
+    public String getAction() {
+        return action;
+    }
+
+    public String getTargetID() {
+        return targetID;
+    }
+
+    public Optional<TEActionReward> getReward() { return Optional.ofNullable(reward); }
+
+    public Map<String, TEActionReward> getRewards() {
+        return rewards;
+    }
+
+    public boolean isValid() {
+        return isValid;
+    }
+
+    public Optional<TEActionReward> evaluateBreak(Logger logger, BlockState state, UUID blockCreator) {
+        // Disqualifying checks first for performance
+        if (!state.getType().getId().equals(this.targetID))
+            return Optional.empty();
+        if (growthTrait == null && blockCreator != null)
+            // A player placed the block and it doesn't indicate growth -> Do not pay to prevent exploits
+            return Optional.empty();
+
+        // Determine base reward - Use complicated way if this block has an ID trait
+        TEActionReward reward = null;
+        if (idTrait != null) {
+            Optional<BlockTrait<?>> trait = state.getTrait(idTrait);
+            if (trait.isPresent()) {
+                Optional<?> traitVal = state.getTraitValue(trait.get());
+                if (traitVal.isPresent()) {
+                    String key = traitVal.get().toString();
+                    reward = rewards.getOrDefault(key, null);
+                }
+            } else {
+                logger.warn("ID trait \"", idTrait, "\" not found during action: ", action, ':', targetID);
```

Comment:
Change single quotes to double quotes for: `':'`

There are a few more spots within this file that need the same change.

Reply:
Why construct a whole new String for just one character that's concatenated later anyways?